File: apps/scala/src/ppl/apps/ml/linreg/LinReg.scala
/* locally weighted linear regression
 *
 * author:  Arvind Sujeeth (asujeeth@stanford.edu)
 *
 * Pervasive Parallelism Laboratory (PPL)
 * Stanford University
 */

package ppl.apps.ml.linreg

import ppl.dsl.optiml._
import ppl.delite.framework.DeliteApplication

object LinRegRunner extends OptiMLApplicationRunner with LinReg

trait LinReg extends OptiMLApplication {

  // unweighted linear regression using the normal equations
  // input: input training vector x
  //        output training vector y
  // output: predictions along uniformly sampled points
  def unweightedReg(x: Rep[DenseVector[Double]], y: Rep[DenseVector[Double]]): Rep[DenseVector[Double]] =
  {
    // by convention, x_0 = 1
    val X = Matrix(Vector.ones(x.length).t, x)

    // theta = inv(X.'X)*(X.'*y) (if y is a col vector)
    val theta = ((X.t*X).inv)*(X.t*y)

    // the resulting fitted line is given by the equation
    //   h(x) = theta_0 + theta_1*x_1 + ...
    theta
  }

  def weightedReg(x: Rep[DenseVector[Double]], y: Rep[DenseVector[Double]]): Rep[DenseVector[Double]] = {
    val tau = 10
    val X = Matrix(Vector.ones(x.length).t, x)

    // initialize prediction points
    val xstep = 25.0/X.numRows
    val xref_pts = Vector.uniform(-10, xstep, 14.99).t
    val xref = Matrix(Vector.ones(xref_pts.length).t, xref_pts)
    //val O = Matrix.identity(X.numRows)
    val Xt = X.t

    // calculate predictions
    val guess = (0::xref.numRows){ e =>
      val x_cur = xref(e,1)
      val weights = x.map(ele => Math.exp(-.1*(x_cur-ele)*(x_cur-ele)/(2.0*tau*tau))/2.0)
      val W = Matrix.diag(weights.length, weights)
      val t1 = Xt*W
      val theta = ((t1*X).inv)*(t1*y)
      (theta.t) *:* (xref(e).t)
    }

    guess
  }

  // file format is m lines with n floats per line, each float seperated by 2 spaces
  // (same as matlab .dat)
  def print_usage = {
    println("Usage: LinRegSerial <input vector file> <output vector file>")
    exit(-1)
  }

  def main() = {
    if (args.length < 2) print_usage

    val x = readVector(args(0)).t
    val y = readVector(args(1)).t

//    logElapsed("Input Section Complete")

    val theta = unweightedReg(x, y)
    println("Unweighted linear regression")
    println("theta: ")
    theta.pprint
    print("\\n")

    tic()
    val guess = weightedReg(x, y)
    toc(guess)

    println("Locally weighted linear regression")
    println("guess: ")
    guess.pprint
    print("\\n")

    //PerformanceTimer.save("LinReg")
  }
}

File: apps/cpp/ml/linreg/linreg.cpp
#include <iostream>
#include <stdlib.h>
#include <math.h>
#include <limits>
#include <ml/OptiML.hpp>
#include "armadillo"

using std::cout;
using namespace arma;

void print_usage() {
  cout << "Usage: LinReg <input vector file> <output vector file>\n";
  exit(-1);
}

int main(int argc,char *argv[]) {

  if(argc != 3) {
    print_usage();
  }

  vec x = OptiML::armad::MLInputReader::readDoubleVectorAsCol(argv[1]);
  vec y = OptiML::armad::MLInputReader::readDoubleVectorAsCol(argv[2]);

  OptiML::tic();
  int tau = 10;
  mat X(x.n_rows,2);
  vec ones(x.n_rows);
  ones.ones();
  OptiML::armad::Matrix::insertColIntoMatrix(0,ones,X);
  OptiML::armad::Matrix::insertColIntoMatrix(1,x,X);

  double xstep = 25.0 / X.n_rows;
  vec xref_pts = OptiML::armad::Vector::uniformCol(-10,xstep,14.99);
  mat xref(xref_pts.n_rows, 2);
  OptiML::armad::Matrix::insertColIntoMatrix(0,ones,xref);
  OptiML::armad::Matrix::insertColIntoMatrix(1,xref_pts,xref);
  mat Xt = trans(X);  

  rowvec guess(xref.n_rows);
  for(uint e = 0; e < xref.n_rows; e++) {
    double x_cur = xref(e,1);
    vec weights(x.n_rows);
    for(uint i = 0; i < x.n_rows; i++) {
      weights(i) = exp(-0.1*pow(x_cur-x(i),2)/(2.0*pow(tau,2)));
    }
    mat W = OptiML::armad::Matrix::doubleMatrixDiag(weights.n_rows, weights);
    mat t1 = Xt*W;
    vec theta = inv(t1*X)*(t1*y);
    guess.at(e) = dot(trans(theta),(trans(xref.row(e))));
  }
  
  OptiML::toc();
  //guess.print("guess = ");
}

